File: bindings/go/metadata.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package opendal

import (
	"context"
	"time"
	"unsafe"

	"github.com/jupiterrider/ffi"
)

// EntryMode indicates whether an entry is a file, directory, or unknown.
type EntryMode uint8

const (
	EntryModeUnknown EntryMode = iota
	EntryModeFile
	EntryModeDir
)

// Metadata represents essential information about a file or directory.
//
// This struct contains attributes commonly used in file systems
// and object storage systems.
type Metadata struct {
	cacheControl       *string
	contentDisposition *string
	contentEncoding    *string
	contentLength      uint64
	contentMD5         *string
	contentType        *string
	etag               *string
	isCurrent          *bool
	isDeleted          bool
	isFile             bool
	isDir              bool
	lastModified       time.Time
	mode               EntryMode
	version            *string
	userMetadata       map[string]string
}

func optionalStringValue(v *string) (string, bool) {
	if v == nil {
		return "", false
	}
	return *v, true
}

func boolPtrFromOptionalByte(v uint8) *bool {
	switch v {
	case 0:
		b := false
		return &b
	case 1:
		b := true
		return &b
	default:
		return nil
	}
}

func newMetadata(ctx context.Context, inner *opendalMetadata) *Metadata {

	var lastModified time.Time
	ms := ffiMetaLastModified.symbol(ctx)(inner)
	if ms != -1 {
		lastModified = time.UnixMilli(ms)
	}

	isCurrent := boolPtrFromOptionalByte(ffiMetaIsCurrent.symbol(ctx)(inner))

	defer ffiMetadataFree.symbol(ctx)(inner)

	return &Metadata{
		cacheControl:       ffiMetaCacheControl.symbol(ctx)(inner),
		contentDisposition: ffiMetaContentDisposition.symbol(ctx)(inner),
		contentEncoding:    ffiMetaContentEncoding.symbol(ctx)(inner),
		contentLength:      ffiMetaContentLength.symbol(ctx)(inner),
		contentMD5:         ffiMetaContentMD5.symbol(ctx)(inner),
		contentType:        ffiMetaContentType.symbol(ctx)(inner),
		etag:               ffiMetaEtag.symbol(ctx)(inner),
		isCurrent:          isCurrent,
		isDeleted:          ffiMetaIsDeleted.symbol(ctx)(inner),
		isFile:             ffiMetaIsFile.symbol(ctx)(inner),
		isDir:              ffiMetaIsDir.symbol(ctx)(inner),
		lastModified:       lastModified,
		mode:               ffiMetaMode.symbol(ctx)(inner),
		version:            ffiMetaVersion.symbol(ctx)(inner),
		userMetadata:       ffiMetaUserMetadata.symbol(ctx)(inner),
	}
}

// CacheControl returns the cache control of the entry.
func (m *Metadata) CacheControl() (string, bool) {
	return optionalStringValue(m.cacheControl)
}

// ContentDisposition returns the content disposition of the entry.
func (m *Metadata) ContentDisposition() (string, bool) {
	return optionalStringValue(m.contentDisposition)
}

// ContentEncoding returns the content encoding of the entry.
func (m *Metadata) ContentEncoding() (string, bool) {
	return optionalStringValue(m.contentEncoding)
}

// ContentMD5 returns the content MD5 of the entry.
func (m *Metadata) ContentMD5() (string, bool) {
	return optionalStringValue(m.contentMD5)
}

// ContentType returns the content type of the entry.
func (m *Metadata) ContentType() (string, bool) {
	return optionalStringValue(m.contentType)
}

// ETag returns the ETag of the entry.
func (m *Metadata) ETag() (string, bool) {
	return optionalStringValue(m.etag)
}

// ContentLength returns the size of the file in bytes.
//
// For directories, this value may not be meaningful and could be zero.
func (m *Metadata) ContentLength() uint64 {
	return m.contentLength
}

// IsFile returns true if the metadata represents a file, false otherwise.
func (m *Metadata) IsFile() bool {
	return m.isFile
}

// IsDir returns true if the metadata represents a directory, false otherwise.
func (m *Metadata) IsDir() bool {
	return m.isDir
}

// IsCurrent returns whether this metadata represents the current version.
//
// The second return value is false when the service doesn't provide this information.
func (m *Metadata) IsCurrent() (bool, bool) {
	if m.isCurrent == nil {
		return false, false
	}
	return *m.isCurrent, true
}

// IsDeleted returns whether this metadata represents a deleted object or version.
func (m *Metadata) IsDeleted() bool {
	return m.isDeleted
}

// LastModified returns the time when the file or directory was last modified.
//
// The returned time is in UTC.
func (m *Metadata) LastModified() time.Time {
	return m.lastModified
}

// Mode returns the mode of the entry.
func (m *Metadata) Mode() EntryMode {
	return m.mode
}

// Version returns the version of the entry.
func (m *Metadata) Version() (string, bool) {
	return optionalStringValue(m.version)
}

// UserMetadata returns the user-defined metadata of the entry.
func (m *Metadata) UserMetadata() map[string]string {
	if m.userMetadata == nil {
		return nil
	}

	metadata := make(map[string]string, len(m.userMetadata))
	for key, value := range m.userMetadata {
		metadata[key] = value
	}
	return metadata
}

var ffiMetaMode = newFFI(ffiOpts{
	sym:    "opendal_metadata_mode",
	rType:  &ffi.TypeUint8,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) EntryMode {
	return func(m *opendalMetadata) EntryMode {
		// libffi may write integer return values using a full register-sized slot
		// even when the declared C return type is u8/bool. Use word-sized
		// storage here, then narrow after the call, to avoid overwriting nearby
		// Go stack memory.
		var mode uint64
		ffiCall(
			unsafe.Pointer(&mode),
			unsafe.Pointer(&m),
		)
		return EntryMode(uint8(mode))
	}
})

var ffiMetaContentLength = newFFI(ffiOpts{
	sym:    "opendal_metadata_content_length",
	rType:  &ffi.TypeUint64,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) uint64 {
	return func(m *opendalMetadata) uint64 {
		var length uint64
		ffiCall(
			unsafe.Pointer(&length),
			unsafe.Pointer(&m),
		)
		return length
	}
})

var ffiMetaCacheControl = newFFIMetadataString("opendal_metadata_cache_control")

var ffiMetaContentDisposition = newFFIMetadataString("opendal_metadata_content_disposition")

var ffiMetaContentMD5 = newFFIMetadataString("opendal_metadata_content_md5")

var ffiMetaContentType = newFFIMetadataString("opendal_metadata_content_type")

var ffiMetaContentEncoding = newFFIMetadataString("opendal_metadata_content_encoding")

var ffiMetaEtag = newFFIMetadataString("opendal_metadata_etag")

var ffiMetaVersion = newFFIMetadataString("opendal_metadata_version")

var ffiMetaIsFile = newFFI(ffiOpts{
	sym:    "opendal_metadata_is_file",
	rType:  &ffi.TypeUint8,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) bool {
	return func(m *opendalMetadata) bool {
		// See ffiMetaMode for why this uses word-sized return storage.
		var result uint64
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&m),
		)
		return uint8(result) == 1
	}
})

var ffiMetaIsDir = newFFI(ffiOpts{
	sym:    "opendal_metadata_is_dir",
	rType:  &ffi.TypeUint8,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) bool {
	return func(m *opendalMetadata) bool {
		// See ffiMetaMode for why this uses word-sized return storage.
		var result uint64
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&m),
		)
		return uint8(result) == 1
	}
})

var ffiMetaIsCurrent = newFFI(ffiOpts{
	sym:    "opendal_metadata_is_current",
	rType:  &ffi.TypeUint8,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) uint8 {
	return func(m *opendalMetadata) uint8 {
		// See ffiMetaMode for why this uses word-sized return storage.
		var result uint64
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&m),
		)
		return uint8(result)
	}
})

var ffiMetaIsDeleted = newFFI(ffiOpts{
	sym:    "opendal_metadata_is_deleted",
	rType:  &ffi.TypeUint8,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) bool {
	return func(m *opendalMetadata) bool {
		// See ffiMetaMode for why this uses word-sized return storage.
		var result uint64
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&m),
		)
		return uint8(result) == 1
	}
})

var ffiMetaLastModified = newFFI(ffiOpts{
	sym:    "opendal_metadata_last_modified_ms",
	rType:  &ffi.TypeSint64,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) int64 {
	return func(m *opendalMetadata) int64 {
		var result int64
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&m),
		)
		return result
	}
})

var ffiMetaUserMetadata = newFFI(ffiOpts{
	sym:    "opendal_metadata_get_user_metadata",
	rType:  &ffi.TypePointer,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) map[string]string {
	return func(m *opendalMetadata) map[string]string {
		var metadata *opendalMetadataUserMetadata
		ffiCall(
			unsafe.Pointer(&metadata),
			unsafe.Pointer(&m),
		)
		if metadata == nil {
			return nil
		}
		defer ffiMetaUserMetadataFree.symbol(ctx)(metadata)

		length := ffiMetaUserMetadataLen.symbol(ctx)(metadata)
		if length == 0 {
			return map[string]string{}
		}

		pairsPtr := ffiMetaUserMetadataPairs.symbol(ctx)(metadata)
		if pairsPtr == nil {
			return map[string]string{}
		}

		pairs := unsafe.Slice(pairsPtr, int(length))
		result := make(map[string]string, len(pairs))
		for _, pair := range pairs {
			key := BytePtrToString(pair.key)
			value := BytePtrToString(pair.value)
			result[key] = value
		}
		return result
	}
})

var ffiMetaUserMetadataPairs = newFFI(ffiOpts{
	sym:    "opendal_metadata_user_metadata_pairs",
	rType:  &ffi.TypePointer,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadataUserMetadata) *opendalMetadataUserMetadataPair {
	return func(m *opendalMetadataUserMetadata) *opendalMetadataUserMetadataPair {
		var result *opendalMetadataUserMetadataPair
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&m),
		)
		return result
	}
})

var ffiMetaUserMetadataLen = newFFI(ffiOpts{
	sym:    "opendal_metadata_user_metadata_len",
	rType:  &ffi.TypeUint64,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadataUserMetadata) uintptr {
	return func(m *opendalMetadataUserMetadata) uintptr {
		var length uint64
		ffiCall(
			unsafe.Pointer(&length),
			unsafe.Pointer(&m),
		)
		return uintptr(length)
	}
})

var ffiMetaUserMetadataFree = newFFI(ffiOpts{
	sym:    "opendal_metadata_user_metadata_free",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadataUserMetadata) {
	return func(m *opendalMetadataUserMetadata) {
		ffiCall(
			nil,
			unsafe.Pointer(&m),
		)
	}
})

func newFFIMetadataString(sym string) *FFI[func(m *opendalMetadata) *string] {
	_ = ffiStringFree
	return newFFI(ffiOpts{
		sym:    contextKey(sym),
		rType:  &ffi.TypePointer,
		aTypes: []*ffi.Type{&ffi.TypePointer},
	}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) *string {
		return func(m *opendalMetadata) *string {
			var bytePtr *byte
			ffiCall(
				unsafe.Pointer(&bytePtr),
				unsafe.Pointer(&m),
			)
			if bytePtr == nil {
				return nil
			}
			value := copyCStringAndFree(bytePtr, ffiStringFree.symbol(ctx))
			return &value
		}
	})
}

var ffiMetadataFree = newFFI(ffiOpts{
	sym:    "opendal_metadata_free",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(m *opendalMetadata) {
	return func(m *opendalMetadata) {
		ffiCall(
			nil,
			unsafe.Pointer(&m),
		)
	}
})

File: core/src/types/metadata.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use std::collections::HashMap;

use chrono::prelude::*;

use crate::raw::*;
use crate::*;

/// Metadata contains all the information related to a specific path.
///
/// Depending on the context of the requests, the metadata for the same path may vary. For example, two
/// versions of the same path might have different content lengths. Keep in mind that metadata is always
/// tied to the given context and is not a global state.
///
/// ## File Versions
///
/// In systems that support versioning, such as AWS S3, the metadata may represent a specific version
/// of a file.
///
/// Users can access [`Metadata::version`] to retrieve the file's version, if available. They can also
/// use [`Metadata::is_current`] and [`Metadata::is_deleted`] to determine whether the metadata
/// corresponds to the latest version or a deleted one.
///
/// The all possible combinations of `is_current` and `is_deleted` are as follows:
///
/// | `is_current`  | `is_deleted` | description                                                                                                                                                                                                                                                                                                                                                                          |
/// |---------------|--------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
/// | `Some(true)`  | `false`      | **The metadata's associated version is the latest, current version.** This is the normal state, indicating that this version is the most up-to-date and accessible version.                                                                                                                                                                                                          |
/// | `Some(true)`  | `true`       | **The metadata's associated version is the latest, deleted version (Latest Delete Marker or Soft Deleted).** This is particularly important in object storage systems like S3. It signifies that this version is the **most recent delete marker**, indicating the object has been deleted. Subsequent GET requests will return 404 errors unless a specific version ID is provided. |
/// | `Some(false)` | `false`      | **The metadata's associated version is neither the latest version nor deleted.** This indicates that this version is a previous version, still accessible by specifying its version ID.                                                                                                                                                                                              |
/// | `Some(false)` | `true`       | **The metadata's associated version is not the latest version and is deleted.** This represents a historical version that has been marked for deletion. Users will need to specify the version ID to access it, and accessing it may be subject to specific delete marker behavior (e.g., in S3, it might not return actual data but a specific delete marker response).             |
/// | `None`        | `false`      | **The metadata's associated file is not deleted, but its version status is either unknown or it is not the latest version.** This likely indicates that versioning is not enabled for this file, or versioning information is unavailable.                                                                                                                                           |
/// | `None`        | `true`       | **The metadata's associated file is deleted, but its version status is either unknown or it is not the latest version.** This typically means the file was deleted without versioning enabled, or its versioning information is unavailable. This may represent an actual data deletion operation rather than an S3 delete marker.                                                   |
#[derive(Debug, Clone, Eq, PartialEq, Default)]
pub struct Metadata {
    mode: EntryMode,

    is_current: Option<bool>,
    is_deleted: bool,

    cache_control: Option<String>,
    content_disposition: Option<String>,
    content_length: Option<u64>,
    content_md5: Option<String>,
    content_range: Option<BytesContentRange>,
    content_type: Option<String>,
    content_encoding: Option<String>,
    etag: Option<String>,
    last_modified: Option<DateTime<Utc>>,
    version: Option<String>,

    user_metadata: Option<HashMap<String, String>>,
}

impl Metadata {
    /// Create a new metadata
    pub fn new(mode: EntryMode) -> Self {
        Self {
            mode,

            is_current: None,
            is_deleted: false,

            cache_control: None,
            content_length: None,
            content_md5: None,
            content_type: None,
            content_encoding: None,
            content_range: None,
            last_modified: None,
            etag: None,
            content_disposition: None,
            version: None,
            user_metadata: None,
        }
    }

    /// mode represent this entry's mode.
    pub fn mode(&self) -> EntryMode {
        self.mode
    }

    /// Set mode for entry.
    pub fn set_mode(&mut self, v: EntryMode) -> &mut Self {
        self.mode = v;
        self
    }

    /// Set mode for entry.
    pub fn with_mode(mut self, v: EntryMode) -> Self {
        self.mode = v;
        self
    }

    /// Returns `true` if this metadata is for a file.
    pub fn is_file(&self) -> bool {
        matches!(self.mode, EntryMode::FILE)
    }

    /// Returns `true` if this metadata is for a directory.
    pub fn is_dir(&self) -> bool {
        matches!(self.mode, EntryMode::DIR)
    }

    /// Checks whether the metadata corresponds to the most recent version of the file.
    ///
    /// This function is particularly useful when working with versioned objects,
    /// such as those stored in systems like AWS S3 with versioning enabled. It helps
    /// determine if the retrieved metadata represents the current state of the file
    /// or an older version.
    ///
    /// Refer to docs in [`Metadata`] for more information about file versions.
    ///
    /// # Return Value
    ///
    /// The function returns an `Option<bool>` which can have the following values:
    ///
    /// - `Some(true)`:  Indicates that the metadata **is** associated with the latest version of the file.
    ///   The metadata is current and reflects the most up-to-date state.
    /// - `Some(false)`: Indicates that the metadata **is not** associated with the latest version of the file.
    ///   The metadata belongs to an older version, and there might be a more recent version available.
    /// - `None`:      Indicates that the currency of the metadata **cannot be determined**. This might occur if
    ///   versioning is not supported or enabled, or if there is insufficient information to ascertain the version status.
    pub fn is_current(&self) -> Option<bool> {
        self.is_current
    }

    /// Set the `is_current` status of this entry.
    ///
    /// By default, this value will be `None`. Please avoid using this API if it's unclear whether the entry is current.
    /// Set it to `true` if it is known to be the latest; otherwise, set it to `false`.
    pub fn set_is_current(&mut self, is_current: bool) -> &mut Self {
        self.is_current = Some(is_current);
        self
    }

    /// Set the `is_current` status of this entry.
    ///
    /// By default, this value will be `None`. Please avoid using this API if it's unclear whether the entry is current.
    /// Set it to `true` if it is known to be the latest; otherwise, set it to `false`.
    pub fn with_is_current(mut self, is_current: Option<bool>) -> Self {
        self.is_current = is_current;
        self
    }

    /// Checks if the file (or version) associated with this metadata has been deleted.
    ///
    /// This function returns `true` if the file represented by this metadata has been marked for
    /// deletion or has been permanently deleted.
    /// It returns `false` otherwise, indicating that the file (or version) is still present and accessible.
    ///
    /// Refer to docs in [`Metadata`] for more information about file versions.
    ///
    /// # Returns
    ///
    /// `bool`: `true` if the object is considered deleted, `false` otherwise.
    pub fn is_deleted(&self) -> bool {
        self.is_deleted
    }

    /// Set the deleted status of this entry.
    pub fn set_is_deleted(&mut self, v: bool) -> &mut Self {
        self.is_deleted = v;
        self
    }

    /// Set the deleted status of this entry.
    pub fn with_is_deleted(mut self, v: bool) -> Self {
        self.is_deleted = v;
        self
    }

    /// Cache control of this entry.
    ///
    /// Cache-Control is defined by [RFC 7234](https://httpwg.org/specs/rfc7234.html#header.cache-control)
    /// Refer to [MDN Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) for more information.
    pub fn cache_control(&self) -> Option<&str> {
        self.cache_control.as_deref()
    }

    /// Set cache control of this entry.
    ///
    /// Cache-Control is defined by [RFC 7234](https://httpwg.org/specs/rfc7234.html#header.cache-control)
    /// Refer to [MDN Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) for more information.
    pub fn set_cache_control(&mut self, v: &str) -> &mut Self {
        self.cache_control = Some(v.to_string());
        self
    }

    /// Set cache control of this entry.
    ///
    /// Cache-Control is defined by [RFC 7234](https://httpwg.org/specs/rfc7234.html#header.cache-control)
    /// Refer to [MDN Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) for more information.
    pub fn with_cache_control(mut self, v: String) -> Self {
        self.cache_control = Some(v);
        self
    }

    /// Content length of this entry.
    ///
    /// `Content-Length` is defined by [RFC 7230](https://httpwg.org/specs/rfc7230.html#header.content-length)
    ///
    /// Refer to [MDN Content-Length](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Length) for more information.
    ///
    /// # Returns
    ///
    /// Content length of this entry. It will be `0` if the content length is not set by the storage services.
    pub fn content_length(&self) -> u64 {
        self.content_length.unwrap_or_default()
    }

    /// Set content length of this entry.
    pub fn set_content_length(&mut self, v: u64) -> &mut Self {
        self.content_length = Some(v);
        self
    }

    /// Set content length of this entry.
    pub fn with_content_length(mut self, v: u64) -> Self {
        self.content_length = Some(v);
        self
    }

    /// Content MD5 of this entry.
    ///
    /// Content MD5 is defined by [RFC 2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html).
    /// And removed by [RFC 7231](https://www.rfc-editor.org/rfc/rfc7231).
    ///
    /// OpenDAL will try its best to set this value, but not guarantee this value is the md5 of content.
    pub fn content_md5(&self) -> Option<&str> {
        self.content_md5.as_deref()
    }

    /// Set content MD5 of this entry.
    pub fn set_content_md5(&mut self, v: &str) -> &mut Self {
        self.content_md5 = Some(v.to_string());
        self
    }

    /// Set content MD5 of this entry.
    pub fn with_content_md5(mut self, v: String) -> Self {
        self.content_md5 = Some(v);
        self
    }

    /// Content Type of this entry.
    ///
    /// Content Type is defined by [RFC 9110](https://httpwg.org/specs/rfc9110.html#field.content-type).
    ///
    /// Refer to [MDN Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) for more information.
    pub fn content_type(&self) -> Option<&str> {
        self.content_type.as_deref()
    }

    /// Set Content Type of this entry.
    pub fn set_content_type(&mut self, v: &str) -> &mut Self {
        self.content_type = Some(v.to_string());
        self
    }

    /// Set Content Type of this entry.
    pub fn with_content_type(mut self, v: String) -> Self {
        self.content_type = Some(v);
        self
    }

    /// Content Encoding of this entry.
    ///
    /// Content Encoding is defined by [RFC 7231](https://httpwg.org/specs/rfc7231.html#header.content-encoding)
    ///
    /// Refer to [MDN Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) for more information.
    pub fn content_encoding(&self) -> Option<&str> {
        self.content_encoding.as_deref()
    }

    /// Set Content Encoding of this entry.
    pub fn set_content_encoding(&mut self, v: &str) -> &mut Self {
        self.content_encoding = Some(v.to_string());
        self
    }

    /// Content Range of this entry.
    ///
    /// Content Range is defined by [RFC 9110](https://httpwg.org/specs/rfc9110.html#field.content-range).
    ///
    /// Refer to [MDN Content-Range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Range) for more information.
    pub fn content_range(&self) -> Option<BytesContentRange> {
        self.content_range
    }

    /// Set Content Range of this entry.
    pub fn set_content_range(&mut self, v: BytesContentRange) -> &mut Self {
        self.content_range = Some(v);
        self
    }

    /// Set Content Range of this entry.
    pub fn with_content_range(mut self, v: BytesContentRange) -> Self {
        self.content_range = Some(v);
        self
    }

    /// Last modified of this entry.
    ///
    /// `Last-Modified` is defined by [RFC 7232](https://httpwg.org/specs/rfc7232.html#header.last-modified)
    ///
    /// Refer to [MDN Last-Modified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified) for more information.
    pub fn last_modified(&self) -> Option<DateTime<Utc>> {
        self.last_modified
    }

    /// Set Last modified of this entry.
    pub fn set_last_modified(&mut self, v: DateTime<Utc>) -> &mut Self {
        self.last_modified = Some(v);
        self
    }

    /// Set Last modified of this entry.
    pub fn with_last_modified(mut self, v: DateTime<Utc>) -> Self {
        self.last_modified = Some(v);
        self
    }

    /// ETag of this entry.
    ///
    /// `ETag` is defined by [RFC 7232](https://httpwg.org/specs/rfc7232.html#header.etag)
    ///
    /// Refer to [MDN ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) for more information.
    ///
    /// OpenDAL will return this value AS-IS like the following:
    ///
    /// - `"33a64df551425fcc55e4d42a148795d9f25f89d4"`
    /// - `W/"0815"`
    ///
    /// `"` is part of etag.
    pub fn etag(&self) -> Option<&str> {
        self.etag.as_deref()
    }

    /// Set ETag of this entry.
    pub fn set_etag(&mut self, v: &str) -> &mut Self {
        self.etag = Some(v.to_string());
        self
    }

    /// Set ETag of this entry.
    pub fn with_etag(mut self, v: String) -> Self {
        self.etag = Some(v);
        self
    }

    /// Content-Disposition of this entry
    ///
    /// `Content-Disposition` is defined by [RFC 2616](https://www.rfc-editor/rfcs/2616) and
    /// clarified usage in [RFC 6266](https://www.rfc-editor/6266).
    ///
    /// Refer to [MDN Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) for more information.
    ///
    /// OpenDAL will return this value AS-IS like the following:
    ///
    /// - "inline"
    /// - "attachment"
    /// - "attachment; filename=\"filename.jpg\""
    pub fn content_disposition(&self) -> Option<&str> {
        self.content_disposition.as_deref()
    }

    /// Set Content-Disposition of this entry
    pub fn set_content_disposition(&mut self, v: &str) -> &mut Self {
        self.content_disposition = Some(v.to_string());
        self
    }

    /// Set Content-Disposition of this entry
    pub fn with_content_disposition(mut self, v: String) -> Self {
        self.content_disposition = Some(v);
        self
    }

    /// Retrieves the `version` of the file, if available.
    ///
    /// The version is typically used in systems that support object versioning, such as AWS S3.
    ///
    /// # Returns
    ///
    /// - `Some(&str)`: If the file has a version associated with it,
    ///   this function returns `Some` containing a reference to the version ID string.
    /// - `None`: If the file does not have a version, or if versioning is
    ///   not supported or enabled for the underlying storage system, this function
    ///   returns `None`.
    pub fn version(&self) -> Option<&str> {
        self.version.as_deref()
    }

    /// Set the version of the file
    pub fn set_version(&mut self, v: &str) -> &mut Self {
        self.version = Some(v.to_string());
        self
    }

    /// With the version of the file.
    pub fn with_version(mut self, v: String) -> Self {
        self.version = Some(v);
        self
    }

    /// User defined metadata of this entry
    ///
    /// The prefix of the user defined metadata key(for example: in oss, it's x-oss-meta-)
    /// is remove from the key
    pub fn user_metadata(&self) -> Option<&HashMap<String, String>> {
        self.user_metadata.as_ref()
    }

    /// Set user defined metadata of this entry
    pub fn with_user_metadata(&mut self, data: HashMap<String, String>) -> &mut Self {
        self.user_metadata = Some(data);
        self
    }
}
